Parallelize the shutdown of VMs

nijinashok opened the pull request:
Currently, the shutdown process of VMs can take a very long time for
large environment since the task wait for the VM to go down before
shutting down the next. Here the shutdown is Parallelized and wait
for default timeout of 300 seconds before forcefully powering off
the VMs.

ovirt-infra commented:
Hello contributor, thanks for submitting a PR for this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure, I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically. ( keep in mind this list might be overwritten if the job XML is refreshed, for permanent whitelisting, please follow #3 option )
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organization's projects.

nijinashok commented:
@tiraboschi @machacekondra can you please review?

sandrobonazzola commented:
ci add to whitelist

tiraboschi commented:
Technically this works, but I'm just thinking if triggering the shutdown of a potentially huge amount of VMs all together can be problematic.
Maybe iterating over a set (with a configurable number) of VMs and repeating it untill we have active VMs.

nijinashok committed:
Parallelize the shutdown of VMs

Currently, the shutdown process of VMs can take a very long time for
large environment since the task wait for the VM to go down before
shutting down the next. Here the shutdown is Parallelized and VMs
will be made down in batches configured by vm_shutdown_batch_count
before forcefully powering off the VMs.

nijinashok commented:
@tiraboschi, thank you for reviewing. Modified as per the recommendation.

didib commented:
I do not know this code at all.

I'd suggest an improvement, though - can be in a later patch.
Sometimes you want to stop a set of VMs before another set of VMs. Say you have appvm1 and appvm2, and dbvm1, where both apps use the same db. You want to first shutdown the apps, then the db. We can allow the user to control this by getting a list of search patterns, and then loop over the list, in each round shutdown all the VMs that match the pattern.

nabeelmuqim commented:
@tiraboschi added to master on Mar 6 where he updated to register ovirt_vm_info result due to ansible 2.9, where result has to be explicitly registered. @nijinashok are you going to update this branch and also merge this branch to master? 

I was also wondering if similar project exist which can only start the VMs which were 'up' prior to shutdown. 

I am planning to connect libreNMS via SNMP to monitor power failure and trigger the parallel shutdown and when power is up, bring up ovirt with only VMs that were in 'up' state before. If you know of any existing work that already perform all these steps together or separately, please share.

arachmani commented:
This repo is deprecated and moved to [ovirt-ansible-collection](https://github.com/oVirt/ovirt-ansible-collection).
In case this PR is still required, please adjust it to fit ovirt-ansible-collection repo or open an RFE.
